Release 2.7

mtmk opened the pull request:
This is the main PR for 2.7 release containing the PRs below. Most have them have minor breaking changes which might affect dependent applications and sometimes transient dependencies. Application developers should recompile and test their application more rigorously in a non-production environment before deploying to production.

0216643 Update INatsJsConsumer to return INatsJsMsg (#1004)
ac5fe82 Fix connection state for consume (#959)
67fdec0 Fix Ad-Hoc JSON Serializer to use Default Options (#984)
d313529 fix: handle 408 Requests Pending responses for fetch requests (#973)
1e658ab Fix publish 503 test (#958)
8f5c735 Fix JetStream publish retry defaults (#939)
887ccde Fix kv ttl interface (#909)
72a2e92 chore: rework NatsOpt.Default initialization (#921)
948af5c fix: `NatsJSConsumer` never disposing receive activities (#911)
bedafe0 Fix build warnings (#912)
5abeb7e fix: `NatsInstrumentationOptions.Default` gets reset each time (#907)
0bc89fd #851 Move Serialization Interface into Abstractions (#858)
5a23eb1 Obsolete ReplyAsync method for NatsJsMsg (#839)
7d9cb11 (thompson-tomo/release/2.7) Add support for `Filter` and `Enrich` for OpenTelemetry activities (#859)
0f7166c (thompson-tomo/split-merge) #848 Tweak dependencies (#853)
b09e92c  #863 Fix naming of consumer group attribute (#870)

mtmk commented on src/NATS.Client/Internals/JsonUtils.cs:
```diff
@@ -144,7 +144,8 @@ public static DateTime AsDate(JSONNode node)
             {
-                return DateTime.Parse(node.Value).ToUniversalTime();
+                DateTime dt = DateTime.Parse(node.Value).ToUniversalTime();
+                return dt.Year == 1 ? DateTime.MinValue : dt; // b/c server sends a value that doesn't quite parse to DateTime.MinValue
```
I don't get this one :-(

scottf replied:
DateTime.MinValue.ToUniversalTime() resolves to `1/1/0001 5:00:00 AM`

But for whatever reason the value the server sends `0001-01-01T00:00:00Z` when resolved with 
```
DateTime.Parse(node.Value).ToUniversalTime()
```
ends up with `1/2/0001 12:00:00 AM`

So I decdide that anything with year 1 must have meant min value. I can't imagine real server date data will have a date of year 1.

mtmk reviewed (approved):
LGTM

thompson-tomo committed:
#863 Fix naming of consumer group attribute (#870)

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

mtmk committed:
Merge branch 'main' into release/2.7

thompson-tomo committed:
#848 Tweak dependencies (#853)

* #848 Tweak dependencies

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* Update NATS.Client.Core.csproj

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* Update NATS.Client.Core.csproj

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* Make STJ explicit dependency for JetStream

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* Update NATS.Client.JetStream.csproj

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* Force sdk to 8.0.0

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* Add newer stj for net 6

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

---------

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

mtmk committed:
Merge branch 'main' into release/2.7

aradalvand committed:
Add support for `Filter` and `Enrich` for OpenTelemetry activities (#859)

* Add support for `Filter` and `Enrich` for OpenTelemetry activities

* Make `internal` methods in `internal Telemetry` `public`

* Fix package versions and whatnot

* Remove `TracerProviderBuilderExtensions`

* Include `Deserialize` in the receive activity

* Revert back accidental change

* Add `ParentContext` to `NatsInstrumentationContext`

* Make `GetActivityContext` public to provide the ability to get context activity context

* Make preprocessor directive more accurate

* Revert .csproj formatting

* Move public artifacts out of the `Internal` namespace/folder

* Fix build script

---------

Co-authored-by: Ziya Suzen <ziya@suzen.net>

yeongjonglim committed:
Obsolete ReplyAsync method for NatsJsMsg (#839)

* Obsolete ReplyAsync method

* Obsolete ReplyAsync method on interface

thompson-tomo committed:
#851 Move Serialization Interface into Abstractions (#858)

* Tweaks

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* Add missing dependencies

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* switch to system.memory

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* Added missing using

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

* Reduce csproj file contents

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

---------

Signed-off-by: James Thompson <thompson.tomo@outlook.com>

mtmk committed:
Merge branch 'main' into release/2.7

mtmk committed:
Release 2.7.0-preview.1 (#905)

* #851 Move Serialization Interface into Abstractions (#858)
* Obsolete ReplyAsync method for NatsJsMsg (#839)
* Add support for `Filter` and `Enrich` for OpenTelemetry activities (#859)
* #848 Tweak dependencies (#853)
* #863 Fix naming of consumer group attribute (#870)

aradalvand committed:
fix: `NatsInstrumentationOptions.Default` gets reset each time (#907)

* initialize `NatsInstrumentationOptions.Default` once

* `NatsInstrumentationContext.ActivityContext` doesn't need to be nullable

* no need for `GetActivityContext` on NatsJSMsg

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	version.txt

mtmk committed:
Release 2.7.0-preview.2 (#908)

* fix: `NatsInstrumentationOptions.Default` gets reset each time (#907)
* Also, fixes from main

mtmk committed:
Fix build warnings (#912)

aradalvand committed:
fix: `NatsJSConsumer` never disposing receive activities (#911)

* Fix `NatsJSConsumer` never disposing receive activities

* Use `ActivityEndingMsgReader` to `NatsJSFetch` and `NatsJSOrderedConsume`

* Consolidate activity ending message readers into one class

* Run `dotnet format`

mtmk committed:
Merge branch 'main' into release/2.7

mtmk committed:
Release 2.7.0-preview.3 (#922)

* fix: `NatsJSConsumer` never disposing receive activities (#911)
* Merge from main

regnrat committed:
chore: rework NatsOpt.Default initialization (#921)

* chore: rework NatsOpt.Default initialization

* chore: removed unused using statement

mtmk committed:
Fix kv ttl interface (#909)

KV TTL should only be allowed on Create and Purge

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	version.txt

mtmk committed:
Merge branch 'main' into release/2.7

mtmk committed:
Release 2.7.0-preview.4 (#931)

Fixes merged from main

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	version.txt

mtmk committed:
Release 2.7.0-preview.5 (#944)

Merge from main aligning with 2.6.8 release.

mtmk committed:
Merge branch 'refs/heads/main' into release/2.7

# Conflicts:
#	version.txt

mtmk committed:
Merge branch 'main' into release/2.7

mtmk committed:
Fix JetStream publish retry defaults (#939)

Decision to retry jetstream publish requests should be letft to the
application since it depends on the delivery and durability
requirements of their solution.

mtmk committed:
Merge branch 'main' into release/2.7

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	version.txt

mtmk committed:
Fix publish 503 test (#958)

With 2.7.x we have made the default to not re-publish on failure.

mtmk committed:
Release 2.7.0-preview.6 (#956)

* Fix JetStream publish retry defaults (#939)
* Merge from main

mtmk committed:
Merge branch 'main' into release/2.7

mtmk committed:
Release 2.7.0-preview.7 (#970)

Merged from main

aradalvand committed:
fix: handle 408 Requests Pending responses for fetch requests (#973)

* Handle 408 Requests Pending responses for fetch requests

* Remove redundant `Console.WriteLine` that was put there for testing

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	version.txt

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	version.txt

mtmk committed:
Release 2.7.0-preview.8 (#986)

* fix: handle 408 Requests Pending responses for fetch requests (#973)
* (merge from main) Object store item size fix (#977)

mtmk committed:
Fix Ad-Hoc JSON Serializer to use Default Options (#984)

mtmk committed:
Fix connection state for consume (#959)

* Fix connection state for consume

* Enhance connection state handling with NatsConnectionFailedException

* Improve error handling for connection failures and add support for configurable 503 error thresholds in JetStream consumers

* Add tests for connection failure handling and configurable 503 error thresholds in JetStream consumers

* Fix format

* Fix test

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	tests/NATS.Client.JetStream.Tests/OrderedConsumerTest.cs

mtmk committed:
Merge branch 'main' into release/2.7

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	src/NATS.Client.JetStream/Internal/NatsJSConsume.cs
#	src/NATS.Client.JetStream/NatsJSConsumer.cs
#	src/NATS.Client.JetStream/NatsJSOpts.cs
#	version.txt

mtmk committed:
Update INatsJsConsumer to return INatsJsMsg (#1004)

mtmk committed:
Merge branch 'main' into release/2.7

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	version.txt

Jarema reviewed (approved):
LGTM!

mtmk committed:
Merge branch 'main' into release/2.7

# Conflicts:
#	src/NATS.Client.Core/NATS.Client.Core.csproj
